### DISCUSSION TITLE
page search

### DISCUSSION THREAD

**Discussion by hugocool** [jheddings/notional#21] [Ideas]
First of all, thank you very much for making this library, its amazing.
Its really cool how well it integrates with Pydantic!

I was expecting there to be a method for searching for pages something like:
```
import notional
from notional.records import Page

notion = notional.connect(auth=NOTION_TOKEN)
notion.pages.query(Page).execute()
```

Which would yield all the pages (no databases)

or maybe something like 
notion.pages.list()

that would also allow for fuzzy matching

notion.pages.list(filter='')

Maybe in the ORM it would be nice to do 

notion.pages.query(page.name= ' ' ).execute()

idk what the API should be here since the notion API does not explicitly searches in the page name for as far as I know.

anyway, now I do:

```
from notion_client import Client
from notional.records import Page

def search_notion(NOTION_TOKEN, query=None):
  base_query = dict(
      filter={
    'property':'object',
    'value':'page'
    },
    sort = {
      "direction":"descending",
      "timestamp":"last_edited_time"
    })
  

  if query:
    full_query = {**base_query,**dict(query=query)}
  else:
    full_query = base_query

  notion = Client(auth=NOTION_TOKEN)
  pages_result = notion.search(**full_query).get("results")

  return [Page.parse_obj(page) for page in pages_result if page.title]
```

(which btw, subclassing Page with pydantic BaseModel opens up a really nice set of validation, parsing and serialisation possibilities so cheers for that!).

**Comment by jheddings** [comment]
Thanks for giving `notional` a spin and for your excellent feedback!

**Comment by jheddings** [comment]
I want to make sure I understand what your use case a bit better...

The `query` endpoint is really intended for database searching.  The Notion API doesn't have an endpoint for explicitly searching `pages`, so we might have to get creative on supporting this.

One option (that I have not implemented yet) would be to expose the `search` endpoint and return an iterator of search results.  This would match any item (Page, Database or Block) from Notion.  This would also allow for a general query string that performs fuzzy matching.  You would simply need to loop through results and only grab items that you are interested in.

So, assuming the objects are properly deserialized for you, it might look like:

```python
result = notion.search(query="interesting page")
return [page for page in result if page isinstance(Page)]
```

I'll have to think about how to add sort and query into this, since we already have the `Query` object to help build sorts and filters.

Would that meet your need?

**Comment by hugocool** [comment]
I think exposing the search endpoint would be a fine place to start and see from there how the API should be structured for consistency.

The Notion Search API does allow one to make a query (fuzzy matching based on provided keyword) or to search without a query (which will return everything that is shared with your integration). 
In addition, one can specify a filter and the sorting of the results.

for example:

```
filter={
    'property':'object',
    'value':'page'
    }
```
will only return pages (and no blocks or databases, but will return pages contained in databases).
The search function I mentioned above will, without a query specified, return all the pages shared with the integration, sorted by their last edited date. The sorting is due to:

```
sort = {
      "direction":"descending",
      "timestamp":"last_edited_time"
    })
```
I totally agree with the idea of modeling this library after SQLalchemy, since that is a very mature library. There are also ODM's which aim to do a similar thing for non-relational databases (it seems to me notion is a kind of postgresql/mongodb hybrid), MongoEngine is a great example https://docs.mongoengine.org/guide/querying.html for inspiration.

Anyway, so in keeping with the idea of querying, I would think one would query based on the Object Model.

So Page.query() would return results from the search API.

The complicating factor here is the fact that in SQlalchemy an object refers to a table in the SQL database. However in Notion there is no such thing as a flat table perse, the object is not linked to specific table by definition, there is no unique mapping between your models and the database. 
While your implementation is working around this by retrieving Object models which are linked to the Notion database due to their unique ID.
This makes querying a little weird, if you query for an un-instantiated class, only in the case of the default Page model this should return a defined result (the list of all pages which conform to that model(which are pages which are not stored in a database).

Basically, there is a difference between pages that are stored in a database and pages which are not. Pages stored in a database are somewhat unpredictable since their properties can be changed easily, and therefore might lose their mapping to your Object Model.
Therefore modelling pages as an object, and subsequently querying for them might require modelling your API after MongoEngine, since Notion will be able to search in every property of the stored data. 
https://docs.mongoengine.org/guide/text-indexes.html

anyway, I am collecting some thoughts on this, I would like to contribute some documentation/examples.
One way to do this would be to use NBdev (https://nbdev.fast.ai/tutorial.html), this allows you to create your package using just jupyter notebooks. This has the added benefit of putting all your code, examples documentation and tests in one place.

**Comment by jheddings** [comment]
I've started a branch for this, but it's incomplete.  Maybe we can iterate there and see if it converges on a good solution?

https://github.com/jheddings/notional/tree/search

I'm fine using some of the tools you mentioned, just haven't taken a look yet so this is a quick start ;)

**Comment by hugocool** [comment]
Unfortunately the search branch is not installable:

pip install git+https://github.com/jheddings/notional.git@search
Collecting git+https://github.com/jheddings/notional.git@search
  Cloning https://github.com/jheddings/notional.git (to revision search) to /tmp/pip-req-build-zjf71qgp
  Running command git clone -q https://github.com/jheddings/notional.git /tmp/pip-req-build-zjf71qgp
  Running command git checkout -b search --track origin/search
  Switched to a new branch 'search'
  Branch 'search' set up to track remote branch 'search' from 'origin'.
  Resolved https://github.com/jheddings/notional.git to commit b9169c965035684f8841bae69de3e67d4e7ac1f3
    ERROR: Command errored out with exit status 1:
     command: /home/gitpod/.pyenv/versions/3.8.12/bin/python3 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-zjf71qgp/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-zjf71qgp/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-a75ztz06
         cwd: /tmp/pip-req-build-zjf71qgp/
    Complete output (9 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-zjf71qgp/setup.py", line 4, in <module>
        from notional.version import __version__
      File "/tmp/pip-req-build-zjf71qgp/notional/__init__.py", line 5, in <module>
        from .session import Session
      File "/tmp/pip-req-build-zjf71qgp/notional/session.py", line 5, in <module>
        import notion_client
    ModuleNotFoundError: No module named 'notion_client'
    ----------------------------------------
WARNING: Discarding git+https://github.com/jheddings/notional.git@search. Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.


So I'll have to clone it before testing it

**Comment by jheddings** [comment]
Yeah, I see the issue there...  I had some module dependences in the setup (for version info).  I've removed that so you should be able to install the branch now.
